### PR TITLE
#11014 Two step getIndexList for other entities

### LINKED ISSUE #11014: Two step `getIndexList` for other entities [5]
### Problem Description

getIndexList uses sort distinct on all attributes that are fetched for directory DTOs, which takes more time with more rows to process. The problem gets bigger with more left joins that multiply rows.

### Proposed Change

Improve performance of `getIndexList` methods (by avoiding extensive sorting on all data columns) that fetch data for the following directories that tend to grow heavily over time:

1. `Task`
2. `Sample`
3. `ExternalMessage`
4. `ShareRequest`
5. `Action`
6. `EventGroup`

### Acceptance Criteria

- [x] Performance analysis before/after is documented (with info about given dataset size).
- [x] At least one unit test exists for each getIndexList method.

### Implementation Details

1. Apply this pattern (already given by #11013):
```java
getIndexList {

  // 1. Fetch ids of entries in the window given by limit & offset
  getIndexListIds {
    sortBy
  }

  // 2. Fetch data for the fetched ids
  executeBatched {
    sortBy
  }

  // 3. Post-processing: calculated values, references, collections
  // 4. Pseudonymization
  // 5. Log execution time (DEBUG)
}
```

### Additional Information

- Follow up to #11013
- No other entities are currently considered to where this refactoring is needed.

### PR DESCRIPTION
**Pull request by leventegal-she**
Fixes #11014

Details to be added on the ticket when closing it

**Task**
* Number of tasks: 131680
* Performance before refactoring:
image
* Performance after refactoring:
image

**Sample**
* Number of samples: 819469
* Performance before refactoring:
image
* Performance after refactoring:
image

**ExternalMessage**
* Number of messages: 39424
* Performance before refactoring: 
image
* Performance after refactoring:
image

**Action**
* Number of actions: 262144
* Performance before refactoring:
image
* Performance after refactoring: 
image

**EventGroup**
* Number of event groups: 163840 
* Performance before refactoring:
image
* Performance after refactoring: 
image

**Share requests index list method was not  refactored because it is more performant as it is now.**
* Number of share requests: 112640
* Performance before refactoring:
image
* Performance after refactoring:
image

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: Please find the results at https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11455

File: sormas-backend/src/main/java/de/symeda/sormas/backend/action/ActionService.java [hunk 1]
```diff
@@ -285,17 +291,100 @@ public Predicate buildEventCriteriaFilter(EventCriteria criteria, ActionQueryCon
...
+				cb.lower(eventResponsibleUser.get(User.LAST_NAME)),
+				cb.selectCase()
+					.when(cb.isNotNull(action.get(Action.LAST_MODIFIED_BY)), cb.lower(lastModifiedBy.get(User.LAST_NAME)))
+					.otherwise(cb.lower(creatorUser.get(User.LAST_NAME))),
+				event.get(Event.CHANGE_DATE),
+				event.get(Event.DELETION_REASON),
+				event.get(Event.OTHER_DELETION_REASON));
+
+			cq.where(action.get(Sample.ID).in(batchedIds));
+
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/action/ActionService.java:359]: I think there's a distinct missing here

File: sormas-backend/src/main/java/de/symeda/sormas/backend/event/EventGroupFacadeEjb.java [hunk 2]
```diff
@@ -164,25 +168,57 @@ public List<EventGroupIndexDto> getIndexList(
...
+			eventCountSubquery.groupBy(eventGroupSubQuery.get(EventGroup.ID));
+
+			cq.multiselect(
+				eventGroup.get(EventGroup.UUID),
+				eventGroup.get(EventGroup.NAME),
+				eventGroup.get(EventGroup.CHANGE_DATE),
+				eventCountSubquery.getSelection());
+
+			cq.where(eventGroup.get(EventGroup.ID).in(batchedIds));
+
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/event/EventGroupFacadeEjb.java:197]: Distinct missing here as well

File: sormas-backend/src/main/java/de/symeda/sormas/backend/externalmessage/ExternalMessageFacadeEjb.java [hunk 3]
```diff
@@ -345,33 +348,62 @@ public List<ExternalMessageIndexDto> getIndexList(
...
+				externalMessage.get(ExternalMessage.PERSON_BIRTH_DATE_MM),
+				externalMessage.get(ExternalMessage.PERSON_BIRTH_DATE_DD),
+				externalMessage.get(ExternalMessage.PERSON_POSTAL_CODE),
+				externalMessage.get(ExternalMessage.STATUS),
+				userJoin.get(User.UUID),
+				userJoin.get(User.FIRST_NAME),
+				userJoin.get(User.LAST_NAME));
+
+			cq.where(externalMessage.get(ExternalMessage.ID).in(batchedIds));
+
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/externalmessage/ExternalMessageFacadeEjb.java:380]: See above

File: sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java [hunk 4]
```diff
@@ -173,158 +174,81 @@ public List<Sample> findBy(SampleCriteria criteria, User user) {
...
+			Root<PathogenTest> testCountRoot = testCountSq.from(PathogenTest.class);
+			testCountSq.where(cb.equal(testCountRoot.get(PathogenTest.SAMPLE), sample), cb.isFalse(testCountRoot.get(PathogenTest.DELETED)));
+			testCountSq.select(cb.countDistinct(testCountRoot.get(PathogenTest.ID)));
+			selections.add(testCountSq.getSelection());
+
+			selections.addAll(getJurisdictionSelections(sampleQueryContext));
+			cq.multiselect(selections);
+
+			cq.where(sample.get(Sample.ID).in(batchedIds));
+
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/sample/SampleService.java:244]: See above

File: sormas-backend/src/main/java/de/symeda/sormas/backend/task/TaskFacadeEjb.java [hunk 5]
```diff
@@ -461,119 +463,154 @@ public long count(TaskCriteria taskCriteria) {
...
+					joins.getCaseFacility().get(Facility.UUID),
+					joins.getCaseFacility().get(Facility.NAME),
+					joins.getCasePointOfEntry().get(PointOfEntry.UUID),
+					joins.getCasePointOfEntry().get(PointOfEntry.NAME)));
+
+			selections.addAll(taskService.getJurisdictionSelections(taskQueryContext));
+			cq.multiselect(selections);
+
+			cq.where(task.get(Task.ID).in(batchedIds));
+
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/task/TaskFacadeEjb.java:545]: See above

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

File: sormas-backend/src/main/java/de/symeda/sormas/backend/action/ActionService.java (hunk 1 above)
Comment (leventegal-she) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/action/ActionService.java:359]: That's true

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: Please find the results at https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11455

**Review by MateStrysewske — APPROVED** [review]

**Pull request merged**